src/infra/core/src/remote_alias_resolver_impl.rs

```diff
@@ -0,0 +1,296 @@
+        access_token_resolver: Arc<dyn OdfServerAccessTokenResolver>,
+        remote_alias_reg: Arc<dyn RemoteAliasesRegistry>,
+    ) -> Self {
+        Self {
+            remote_repo_reg,
+            access_token_resolver,
+            remote_alias_reg,
+        }
+    }
+
+    async fn fetch_remote_url(
+        &self,
+        local_handle: &odf::DatasetHandle,
+        remote_alias_kind: RemoteAliasKind,
+    ) -> Result<Option<Url>, ResolveAliasError> {
+        let remote_aliases = self
+            .remote_alias_reg
+            .get_remote_aliases(&local_handle.as_local_ref())
+            .await
+            .int_err()?;
+
+        let aliases: Vec<_> = remote_aliases.get_by_kind(remote_alias_kind).collect();
+
+        match aliases.len() {
+            0 => Ok(None),
+            1 => {
+                if let odf::DatasetRefRemote::Url(remote_url) = aliases[0].clone() {
+                    return Ok(Some(remote_url.as_ref().clone()));
+                }
+                Ok(None)
+            }
+            _ => Err(ResolveAliasError::AmbiguousAlias),
+        }
+    }
+
+    fn combine_remote_url(
+        &self,
+        repo_url: &Url,
+        account_name_maybe: Option<&odf::AccountName>,
+        dataset_name: &odf::DatasetName,
+    ) -> Result<Url, InternalError> {
+        let mut res_url = repo_url.clone().as_odf_protocol().int_err()?;
+        {
+            let mut path_segments = res_url.path_segments_mut().unwrap();
+            if let Some(account_name) = account_name_maybe {
+                path_segments.push(account_name);
+            }
+            path_segments.push(dataset_name);
+        }
+        Ok(res_url)
+    }
+
+    async fn resolve_remote_dataset_name(
+        &self,
+        dataset_handle: &odf::DatasetHandle,
+        remote_repo_url: &Url,
+        access_token_maybe: Option<&String>,
+    ) -> Result<odf::DatasetName, ResolveAliasError> {
+        let result = if let Some(remote_dataset_name) =
+            RemoteAliasResolverApiHelper::fetch_remote_dataset_name(
+                remote_repo_url,
+                &dataset_handle.id,
+                access_token_maybe,
+            )
+            .await?
+        {
+            remote_dataset_name
+        } else {
+            dataset_handle.alias.dataset_name.clone()
+        };
+        Ok(result)
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl RemoteAliasResolver for RemoteAliasResolverImpl {
+    async fn resolve_remote_alias(
+        &self,
+        local_dataset_handle: &odf::DatasetHandle,
+        dataset_push_target_maybe: Option<odf::DatasetPushTarget>,
+        remote_alias_kind: RemoteAliasKind,
+    ) -> Result<RemoteAliasRef, ResolveAliasError> {
+        let repo_name: odf::RepoName;
+        let mut account_name = None;
+        let mut dataset_name = None;
```

Comment:
Might be better to rewrite this as:
```rust
let (repo_name, account_name, dataset_name) = if ... { .. } else { ... };
```
This way you get certainty that all branches of logic consider all three of these values and not simply forget to assign one of them.